flake.nix

```diff
@@ -16,132 +16,123 @@
 
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
-    flake-utils.url = "github:numtide/flake-utils";
+
+    #
+    # Flake and repo structuring configurations
+    #
+    # Allows us to structure the flake with the NixOS module system
     flake-parts = {
       url = "github:hercules-ci/flake-parts";
       inputs.nixpkgs-lib.follows = "nixpkgs";
     };
+
+    flake-root.url = "github:srid/flake-root";
+
+    lib-extras = {
+      url = "github:aldoborrero/lib-extras";
+      inputs.flake-parts.follows = "flake-parts";
+      inputs.flake-root.follows = "flake-root";
+      inputs.nixpkgs.follows = "nixpkgs";
+      inputs.treefmt-nix.follows = "treefmt-nix";
+    };
+
+    # Format all the things
+    treefmt-nix = {
+      url = "github:numtide/treefmt-nix";
+      inputs.nixpkgs.follows = "nixpkgs";
+    };
+
+    # To ensure that checks are run locally to enforce cleanliness
+    pre-commit-hooks-nix = {
+      url = "github:cachix/pre-commit-hooks.nix";
+      inputs = {
+        nixpkgs.follows = "nixpkgs";
+        nixpkgs-stable.follows = "nixpkgs";
+        flake-utils.follows = "flake-utils";
+        flake-compat.follows = "flake-compat";
+      };
+    };
+
+    # For preserving compatibility with non-Flake users
+    flake-compat = {
```

Comment:
I'm not sure is adding flake-compat a good idea? I'm not completely against it but I'm also not going to use it.

Previously when we didn't have changed to declarative configurations with microvm.nix it wouldn't have even been possible to add flake-compat. But now I'm just a bit afraid that it might make it even more confusing for newcomers joining the project

Reply:
the primary aim is so that folks who do nt have nix enabled in their system can just build the solution. All it does is to hide the flake goodness and give folks a standard default.nix and shell.nix experience

Reply:
I agree with Mika. In addition, it adds to complexity and future maintenance promise.

Reply:
I am not sure I understand the added complexity? It is used to make a flake into something that a user can run without having to enable Experimental features. It exists only in the default.nix and shell.nix entry points for those folks. For others the flake and nix3 functions are the starting point. It has no impact in the flake flow. 

Maybe there will be some breaking change that will inhibit the shell.nix from working but in terms of complexity i must have missed something on what flake compat does, other than set the hash of all the repos in a more traditional nix way.

Reply:
> I am not sure I understand the added complexity? It is used to make a flake into something that a user can run without having to enable Experimental features. It exists only in the default.nix and shell.nix entry points for those folks. For others the flake and nix3 functions are the starting point. It has no impact in the flake flow.

At high level it all boils down to this - https://github.com/NixOS/nix/issues/4024 - "Flake is not backward compatible with usual non-flake commands". Nor will it be as there is no such promise with flakes.
 
> Maybe there will be some breaking change that will inhibit the shell.nix from working but in terms of complexity i must have missed something on what flake compat does, other than set the hash of all the repos in a more traditional nix way.

Practically it will likely work for most cases as the flake-compat is written by the nix BDFL themself. For me, the added complexity comes from the question: "Is this something we can't take away?" If the answer is no, it adds to complexity. IMO - sometimes less is more. The practical side of keeping it is that if some decides to use it and finds some esoteric corner case, will anyone of us care to look at it? And when we care, can we help?

Anyway - it might be practical to have it for some people and I don't strongly oppose it. I digress but thanks for asking for clarification.